### PR TITLE
AI action: Ping

### LINKED ISSUE #7: AI Action - Ping

### PR DESCRIPTION
**Pull request by ColinJPage**
The Game object now has a function "scrambleThenPing()". It takes in a row, column, and threat type. Before performing the ping, between 0 and 2 of the fields are randomly scrambled.

I also made a new object for Room so we don't have to represent them as strings.

closes #7

### CODE REVIEW

File: app/game/game.js [hunk 1]
```diff
@@ -1,27 +1,61 @@
...
-    const rooms = avilableRooms.filter((room) => !(room in roomsWithThreats));
-    return rooms[randomInt(0, rooms.length - 1)];
+const selectThreatRoom = (availableRooms, roomsWithThreats) => {
+    const roomCandidates = availableRooms.filter((room) => !(room in roomsWithThreats));
+    return roomCandidates[randomInt(0, roomCandidates.length - 1)];
 }
 
 const GAME = (humanUsername, aiUsername) => {
-    let gameTime = 4*60;
+    const gameTime = 4*60;
```
Comment (RichardLam772001) [inline app/game/game.js:20]: If we're using timeRemaining as the decrementing time, we don't need gameTime here. It just gets assigned to timeRemaining.

File: app/game/game.js [hunk 2]
```diff
@@ -32,40 +66,130 @@ const GAME = (humanUsername, aiUsername) => {
...
+    //Retuns string representation
     const getCurrentRoom = () => {
-        return room;
+        return `${humanRoom.x}-${humanRoom.y}`;
+    }
+    const trySpawnThreat= () => {
+        if(THREATS.length >= MAX_THREATS_ACTIVE) return; //too many threats
+
+        const threatRoom = selectThreatRoom(AVAILABLE_ROOMS, ROOMS_WITH_THREATS);
+        if(threatRoom == undefined) return; // No valid room was picked
```
Comment (RichardLam772001) [inline app/game/game.js:99]: Nitpick, this can be replaced with
```
if (!threatRoom) {
       return
}
```

File: app/game/game.js [hunk 3]
```diff
@@ -32,40 +66,130 @@ const GAME = (humanUsername, aiUsername) => {
...
+            return onTimeUp();
+        }
     }
+    
     const onThreatUnresolved = (room, threatId) => {
         THREATS.splice(threatId, 1); // Remove threat from THREATS list
         AVAILABLE_ROOMS.splice(AVAILABLE_ROOMS.indexOf(room), 1);
-        console.log(`Threat was unresolved room ${room} is no longer available`);
+        let message = `ROOM ${room.name} HAS BEEN DESTROYED BY ${"<threatType>"}`;
+        let line = ConsoleLineData(timeRemaining, message, undefined, "critical");
```
Comment (RichardLam772001) [inline app/game/game.js:126]: Nitpick, message and time are not reassigned, these should be const
```
let message = `ROOM ${room.name} HAS BEEN DESTROYED BY ${"<threatType>"}`;
let line = ConsoleLineData(timeRemaining, message, undefined, "critical");
```

File: app/game/game.js [hunk 4]
```diff
@@ -32,40 +66,130 @@ const GAME = (humanUsername, aiUsername) => {
...
+    const roomIsDestroyed = (room) =>{
+        return AVAILABLE_ROOMS.indexOf(room) === -1;
+    }
+    const enterRoom = (roomString) => { //TODO: Validation
+        let coords = roomString.split("-");
+        return enterRoomPos(coords[0], coords[1]);
+    }
+    const enterRoomPos = (x,y) => {
+        if(!validateRoomPos(x,y)) return;
+        let room = ROOMS[x][y];
```
Comment (RichardLam772001) [inline app/game/game.js:82]: nitpick, room is not reassigned so should be const
```
const room = ROOMS[x][y];
```

File: app/game/game.js [hunk 5]
```diff
@@ -32,40 +66,130 @@ const GAME = (humanUsername, aiUsername) => {
...
-            console.log(`GAME - Cannot enter room ${newRoom}`);
+    //Whether the room exists on the spaceship
+    const validateRoomPos = (x, y) =>{
+        return x >= 0 && x < SHIP_ROWS && y >= 0 && y < SHIP_COLS;
+    }
+    const roomIsDestroyed = (room) =>{
+        return AVAILABLE_ROOMS.indexOf(room) === -1;
+    }
+    const enterRoom = (roomString) => { //TODO: Validation
+        let coords = roomString.split("-");
```
Comment (RichardLam772001) [inline app/game/game.js:77]: nitpick, coords also should be const
```
const coords = roomString.split("-");
```

File: app/game/game.js [hunk 6]
```diff
@@ -32,40 +66,130 @@ const GAME = (humanUsername, aiUsername) => {
         }
         return null;
     }
-    const enterRoom = (newRoom) => {
-        // If room not available (cause destroyed), cant enter
-        if (AVAILABLE_ROOMS.indexOf(newRoom) === -1) {
-            console.log(`GAME - Cannot enter room ${newRoom}`);
+    //Whether the room exists on the spaceship
+    const validateRoomPos = (x, y) =>{
+        return x >= 0 && x < SHIP_ROWS && y >= 0 && y < SHIP_COLS;
```
Comment (RichardLam772001) [inline app/game/game.js:71]: Ah, I used to do game development during the first 2 years of programming too, this is typical of position validation in gaming. Haven't seen it in web dev much lol

File: app/game/game.js [hunk 7]
```diff
@@ -1,27 +1,61 @@
 const { ROLES, GAME_TICK_DELAY_MS } = require("../constants");
-const { Threat, THREAT_COOLDOWN_SECONDS } = require("./threat");
-const { randomInt } = require("../utils.js");
+const { Threat, THREAT_COOLDOWN_SECONDS, MAX_THREATS_ACTIVE } = require("./threat");
+const { randomInt, randomChoice } = require("../utils.js");
```
Comment (RichardLam772001) [inline app/game/game.js:3]: randomChoice isn't used anywhere, we should remove unused import
```
const { randomInt } = require("../utils.js");
```

**Review by RichardLam772001 — CHANGES_REQUESTED** [review]
Good work! I see you've added some ping scrambling and refactored a few things. Some small nitpick changes requested.
Also in terms of testing:
scrambleThenPing isn't used anywhere currently but I know it'll probably be used in the future. I don't usually like merging in something I can't test easily, but since overall changes are small I think we're safe letting it in

File: app/game/game.js (hunk 1 above)
Comment (ColinJPage) [reply app/game/game.js:20]: This makes sense. We may eventually need to track both the remaining time and the total time, like if we had a big long progress bar at the top of the screen representing the time until rescue. But I'll consolidate this for now

**Review by RichardLam772001 — APPROVED** [review]

**Pull request merged**